File: programmers/Greedy/SpeedCamera.java
package programmers.Greedy;

import java.util.Arrays;

class SpeedCamera {
    public int solution(int[][] routes) {
        Arrays.sort(routes, (o1, o2) -> o1[0] != o2[0] ? o1[0] - o2[0] : o1[1] - o2[1]);
        int answer = 1;
        int top = routes[0][1];
        int bottom = routes[0][0];
        for (int i = 1; i < routes.length; i++) {
            if (bottom <= routes[i][0] && routes[i][0] <= top) {
                top = Math.min(top, routes[i][1]);
                bottom = routes[i][0];
            } else  {
                top = routes[i][1];
                bottom = routes[i][0];
                answer++;
            }
        }
        return answer;
    }
}

class Main {
    public static void main(String[] args) {
        int[][] routes = {{-20, -15}, {-14, -5}, {-18, -13}, {-5, -3}};
        int answer = new SpeedCamera().solution(routes);
        System.out.print(answer);
    }
}

File: programmers/Greedy/SpeedCamera.kt
package programmers.Greedy

class KSpeedCamera {
    fun solution(routes: Array<IntArray>): Int {
        routes.sortWith(compareBy({ it[0] }, { it[1] }))
        var answer = 1
        var top = 30_000
        var bottom = -30_000
        for ((from, to) in routes) {
            if (from in bottom..top) {
                top = minOf(top, to)
                bottom = from
            } else {
                top = to
                bottom = from
                answer++
            }
        }
        return answer
    }
}

fun main() {
    val routes = arrayOf(
        intArrayOf(-20, -15),
        intArrayOf(-14, -5),
        intArrayOf(-18, -13),
        intArrayOf(-5, -3)
    )
    val answer = KSpeedCamera().solution(routes)
    print(answer)
}